chain/epoch_manager/src/reward_calculator.rs

```diff
@@ -83,47 +88,127 @@ impl RewardCalculator {
         }
         let epoch_validator_reward = epoch_total_reward - epoch_protocol_treasury;
         let mut epoch_actual_reward = epoch_protocol_treasury;
-        let total_stake: Balance = validator_stake.values().sum();
-        for (account_id, stats) in validator_block_chunk_stats {
-            // Uptime is an average of block produced / expected and chunk produced / expected.
-            let average_produced_numer = U256::from(
-                stats.block_stats.produced * stats.chunk_stats.expected
-                    + stats.chunk_stats.produced * stats.block_stats.expected,
-            );
-            let average_produced_denom =
-                U256::from(2 * stats.chunk_stats.expected * stats.block_stats.expected);
-            let online_min_numer = U256::from(*self.online_min_threshold.numer() as u64);
-            let online_min_denom = U256::from(*self.online_min_threshold.denom() as u64);
-            // If average of produced blocks below online min threshold, validator gets 0 reward.
-            let reward = if average_produced_numer * online_min_denom
-                < online_min_numer * average_produced_denom
-                || stats.chunk_stats.expected == 0
-                || stats.block_stats.expected == 0
-            {
-                0
-            } else {
+
+        if checked_feature!(
+            "protocol_feature_chunk_only_producers",
+            ChunkOnlyProducers,
+            protocol_version
+        ) {
+            let to_big_rational = |rational: Rational| -> BigRational {
+                BigRational::new((*rational.numer()).into(), (*rational.denom()).into())
+            };
+            let online_min_threshold = to_big_rational(self.online_min_threshold);
+            let online_max_threshold = to_big_rational(self.online_max_threshold);
+            let calc_uptime_mult = |stats: &ValidatorStats| -> BigRational {
+                if stats.expected != 0 {
+                    let produced_rate =
+                        BigRational::new(stats.produced.into(), stats.expected.into());
+                    let uptime_dividend = produced_rate - &online_min_threshold;
+                    let uptime_divisor = &online_max_threshold - &online_min_threshold;
+                    max(
+                        min(uptime_dividend / uptime_divisor, BigRational::from_integer(1.into())),
+                        BigRational::from_integer(0.into()),
+                    )
+                } else {
+                    BigRational::from_integer(0.into())
+                }
+            };
+            let bp_total_stake: Balance = validator_stake
+                .iter()
+                .map(|(account_id, stake)| {
+                    if validator_block_chunk_stats
+                        .get(account_id)
+                        .unwrap_or_else(|| panic!("{} is not a validator", account_id))
```

Comment:
this might be not correct..

as validator_block_chunk_stats contains only validators that were not kicked out in this epoch.
So if validator did get a kickout, it will not longer be in this list, and we might panic.